### PR TITLE
Cluster-autoscaler: introduce ClusterStateRegistry to main

### PR DESCRIPTION
**Pull request by mwielgus**
Ref: #2228 #2229 

cc: @jszczepkowski @fgrzadkowski @piosz

### CODE REVIEW

File: cluster-autoscaler/scale_down.go [hunk 1]
```diff
@@ -50,6 +50,13 @@ const (
 	ScaleDownNodeDeleted ScaleDownResult = iota
 )
 
+const (
+	// MaxKubernetesNodeDeletiontime is the maximum time needed by Kubernetes to delete an empty node.
+	MaxKubernetesEmptyNodeDeletiontime = 3 * time.Minute
```
Comment (jszczepkowski) [inline cluster-autoscaler/scale_down.go]: Do you intend to make it configurable by flag in future?

File: cluster-autoscaler/scale_down.go [hunk 2]
```diff
@@ -197,14 +204,28 @@ func (sd *ScaleDown) TryToScaleDown(nodes []*apiv1.Node, pods []*apiv1.Pod) (Sca
...
+				finalError = fmt.Errorf("Failed to delete nodes in time")
+				break
+			}
+			select {
+			case err := <-confirmation:
+				if err != nil {
+					glog.Errorf("Problem with empty node deletion: %v", err)
+					finalError = err
+				}
+			case <-time.After(timeLeft):
```
Comment (jszczepkowski) [inline cluster-autoscaler/scale_down.go]: Will it behave correctly if `timeLeft` is negative?

Comment (jszczepkowski) [inline cluster-autoscaler/scale_down.go]: Do you have a unittest for timeout here?

File: cluster-autoscaler/scale_down.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/scale_down.go]: Yes

File: cluster-autoscaler/scale_down.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/scale_down.go]: See 10 lines above :).

Comment (jszczepkowski) [inline cluster-autoscaler/scale_down.go]: OK ;) how abut the test?

Comment (mwielgus) [inline cluster-autoscaler/scale_down.go]: Not at this moment, I will be adding cloud provider failure related tests later.

**Comment by jszczepkowski** [comment]
LGTM

**Pull request merged**